Add Support of healthcheck in docker_container module

akshay196 opened the pull request:
Now container can be started with healthcheck enabled

##### SUMMARY
Fixes #33622

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_container

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
ansible 2.8.0.dev0 (docker-healthcheck-fix-33622 71d323d402) last updated 2018/10/10 20:45:20 (GMT +550)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/akshay/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/akshay/source/ansible/lib/ansible
  executable location = /home/akshay/source/ansible/bin/ansible
  python version = 3.6.6 (default, Jul 19 2018, 14:25:17) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
Container started with healtcheck enabled using docker_container module.

    - name: Run nginx docker container
      docker_container:
        name: nginx-example1
        image: nginx:1.13
        state: started
        healthcheck:
          test: ["CMD-SHELL", "curl --fail http://nginx.host.com || exit 1"]
          interval: 5m
          timeout: 3s
          retries: 3
          start_period: 0s

```paste below
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                            PORTS               NAMES
841853243dbe        nginx:1.13          "nginx -g 'daemon of…"   2 minutes ago       Up 2 minutes (health: starting)   80/tcp              nginx-example1
```

akshay196 committed:
Add Support of healthcheck in docker_container module

Fixes #33622
Now container can be started with healthcheck enabled

Signed-off-by: Akshay Gaikwad <akgaikwad001@gmail.com>

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -147,6 +147,15 @@
     description:
       - List of additional group names and/or IDs that the container process will run as.
+  healthcheck:
+    type: dict
```
You need to add `version_added: "2.8"`

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -147,6 +147,15 @@
+    type: dict
+    description:
+      - Configure a check that's run to determine whether container is healthy
+      - 'I(test) - Test to perform to determine container health.'
```
You should document the elements in the dict better. It is not clear from this short description that `test` is a list of strings, which is the command which will be executed (and what `CMD` in front of it does). Also, it is unclear to me how this command reports the health (return code? timeout?). If it is too complicated to describe it here, you should link to a document where this is explained in detail (docker documentation?) and say that it is similar to there.

felixfontein replied:
In general: you should also document default values, if there are any.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -147,6 +147,15 @@
+    description:
+      - Configure a check that's run to determine whether container is healthy
+      - 'I(test) - Test to perform to determine container health.'
+      - 'I(interval) - The time to wait between check.'
```
Here (or somewhere else) you should describe in which format time can be specified.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -641,6 +650,18 @@
+    image: image_name
+    state: started
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost"]
```
Maybe you should add a short comment explaining what this test achieves. I.e. something like `Tries to make a HTTP GET to localhost. If this fails or timeouts, the healthcheck fails.` (if that's correct).

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1269,54 @@ def _parse_log_config(self):
+        if self.healthcheck.get('test') is None:
+            return None
+
+        if self.healthcheck.get('interval') is None:
```
Why don't you create a new dict which only contains the parameters you support?

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1269,54 @@ def _parse_log_config(self):
+            return None
+
+        if self.healthcheck.get('interval') is None:
+            self.healthcheck['interval'] = 30000000000  # (default: 30s)
```
The defaults should be documented above.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1269,54 @@ def _parse_log_config(self):
+        Return time duration in nanosecond
+        '''
+        regex = re.compile(r'((?P<hours>\d+?)hr)?((?P<minutes>\d+?)m)?((?P<seconds>\d+?)s)?')
+        parts = regex.match(time_str)
```
You should handle the case that `time_str` is not actually a string.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1269,54 @@ def _parse_log_config(self):
+        parts = regex.match(time_str)
+
+        if not parts:
+            return
```
If you can't parse it, you should fail instead of returning `None`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1269,54 @@ def _parse_log_config(self):
+        '''
+        Return time duration in nanosecond
+        '''
+        regex = re.compile(r'((?P<hours>\d+?)hr)?((?P<minutes>\d+?)m)?((?P<seconds>\d+?)s)?')
```
Why don't you allow fractions of a second? (Not that it makes too much sense, but after all, docker expects nanoseconds.)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -753,6 +775,7 @@ def __init__(self, client):
         self.exposed_ports = None
         self.force_kill = None
         self.groups = None
+        self.healthcheck = dict()
```
Don't use `dict()` here; use `None`. It will be overwritten anyway.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -845,6 +868,10 @@ def __init__(self, client):
         self.log_config = self._parse_log_config()
+
+        if self.healthcheck:
+            self.healthcheck = self._parse_healthcheck_options()
```
Remove the `if` and make `_parse_healthcheck_options()` handle `self.healthcheck == None`.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1269,54 @@ def _parse_log_config(self):
+
+        return self.healthcheck
+
+    def convert_duration_to_nanosecond(self, time_str):
```
This is a private function, so its name should start with a `_`.

felixfontein reviewed (changes requested):
Thanks for the PR!

Looks good. Besides all the little things I commented on, there's one thing which is not good: the option is not checked for idempotency.

If the healthcheck details can only be set on container creation, you need to check `Container.has_different_configuration()`. If the healthcheck details can be changed during runtime, it must be in `Container.has_different_resource_limits()` / `TaskParameters.update_parameters()` / `Container.update_limits()` (in that case, we should rename some of the functions).

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -147,6 +147,21 @@
+      - 'Configure a check that is run to determine whether or not containers for this service are "healthy".
+        See the docs for the L(HEALTHCHECK Dockerfile instruction,https://docs.docker.com/engine/reference/builder/#healthcheck)
+        for details on how healthchecks work.'
+      - 'I(test) - Command to run to check health. C(test) must be either string or list. If it is list first item must
```
`C(test) must be either *a* sting or *a* list. If it is *a* list, *the* first item must be *one of* C(NONE), ...`

akshay196 replied:
My bad. I will update this. Thank you @felixfontein for pointing.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1274,58 @@ def _parse_log_config(self):
+        duration_options = ['interval', 'timeout', 'start_period']
+
+        for (key, value) in supported_healthcheck_options.items():
+            if not self.healthcheck.get(key):
```
You should write `if key not in self.healthcheck:` to avoid confusing falsish values with non-specified values.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1274,58 @@ def _parse_log_config(self):
+                    # Convert provided durations into nanoseconds
+                    self.healthcheck[key] = self._convert_duration_to_nanosecond(self.healthcheck[key])
+
+        return self.healthcheck
```
Again: why don't you create a new dict which only contains the parameters you support?

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -931,6 +963,7 @@ def create_parameters(self):
             labels='labels',
             stop_signal='stop_signal',
             working_dir='working_dir',
+            healthcheck='healthcheck',
```
`healthcheck` was only added in docker-py 2.0 (see [changelog](https://docker-py.readthedocs.io/en/stable/change-log.html)), while we support also older versions of docker-py. You have to make sure that this parameter is only added for docker-py 2.0 or newer (and inform the user that if `healthcheck` is specified and docker-py < 2.0, that the value will be ignored -- see other examples in `docker_container`, like `init`).

felixfontein reviewed (commented):
The idempotency question is still open. Also, there's something you have to think about: how can the user indicate that an existing healthcheck for a container should be removed? There must be three states:
 1. Healthcheck is explicitly added
 2. Healthcheck is explicitly removed (maybe indicate as empty dict?)
 3. Healthcheck is ignored (when parameter `healthcheck` is `None` / not specified)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1241,6 +1276,59 @@ def _parse_log_config(self):
+                time_params[name] = int(value)
+
+        time_in_seconds = timedelta(**time_params).seconds
+        time_in_nanoseconds = time_in_seconds * 1000000000
```
Here you ignore the fractional parts (milliseconds, microseconds). You need to add `1000 * timedelta(**time_params).microseconds`. (Also, you should store `timedelta(**time_params)` in a variable, instead of `time_in_seconds`, and work with that one.)

I.e. something like:
``` .py
time = timedelta(**time_params)
time_in_nanoseconds = (time.seconds * 1000000 + time.microseconds) * 1000
```

felixfontein reviewed (commented):
Looks good! I think the only unsolved problem is: how to remove healthcheck from an existing container? Specifying an empty dict for `healthcheck` does not remove it (since it is interpreted the same as when it is not specified).

felixfontein commented:
Two more things:
 1) Can you add a small changelog fragment? You can take this one as a template: https://github.com/ansible/ansible/blob/devel/changelogs/fragments/44789-docker_container-comparisons.yaml
 2) Can you add a short integration test? Maybe similar to this one: https://github.com/ansible/ansible/blob/devel/test/integration/targets/docker_container/tasks/tests/options.yml#L1436-L1495

As soon as you added a way to disable healthchecks, you should also add that to the test (i.e. the sequence should be enable, idempotency, change, disable) :)

felixfontein commented:
Regarding implementing disabling: maybe you should define `parameters.disable_healthcheck` as a boolean which is set to `True` if `self.healthcheck` is not `None` but empty (before `self.healthcheck = self._parse_healthcheck()`), and to `None` otherwise.

Then you can add `disable_healthcheck` to `config_mapping` in `has_different_configuration()` and set it to `True` if `config.get('Healthcheck')` is `None` or empty. That way (since `parameters.disable_healthcheck` is either `True` or `None`), idempotency handling should work as expected.

felixfontein commented:
I've created a PR against your branch with some of the missing things: akshay196/ansible#1

akshay196 commented:
@felixfontein Thank for your help. I appreciate it.
(1)
> I think the only unsolved problem is: how to remove healthcheck from an existing container? Specifying an empty dict for healthcheck does not remove it (since it is interpreted the same as when it is not specified).

As mentioned in Docker docs[1] and docker-py docs[2], we can disable/remove healthcheck by specifying `test: ["NONE"]` . I guess, It could be better if we add this disable method in the module documentation as an example.
As this way is available, do we need to implement disabling for empty dict?

One more approach is to add `disable: true` to disable healthcheck. (Currently this option is available in healthcheck of docker compose file)


(2)
> Can you add a small changelog fragment?

Sure. I will add change log fragment in next update.

(3)
> Can you add a short integration test?

You have added simple integration test in the PR https://github.com/akshay196/ansible/pull/1
After what we decide about disabling, we can do additional changes in that as require.



 

[1] https://docs.docker.com/compose/compose-file/#healthcheck
[2] https://docker-py.readthedocs.io/en/stable/api.html#docker.types.Healthcheck

felixfontein commented:
> > I think the only unsolved problem is: how to remove healthcheck from an existing container? Specifying an empty dict for healthcheck does not remove it (since it is interpreted the same as when it is not specified).
> 
> As mentioned in Docker docs[1] and docker-py docs[2], we can disable/remove healthcheck by specifying `test: ["NONE"]` . I guess, It could be better if we add this disable method in the module documentation as an example.
> As this way is available, do we need to implement disabling for empty dict?

Sure, that can be done as well. I implemented it here: https://github.com/akshay196/ansible/pull/1/commits/c701a648918336677ff4c62e3ba01a2170f52b46 (I've updated the PR akshay196#1 in general.)

> One more approach is to add `disable: true` to disable healthcheck. (Currently this option is available in healthcheck of docker compose file)

I'd prefer the `test: ['NONE']` way.

> > Can you add a small changelog fragment?
> 
> Sure. I will add change log fragment in next update.

Looks good!

> > Can you add a short integration test?
> 
> You have added simple integration test in the PR [akshay196#1](https://github.com/akshay196/ansible/pull/1)
> After what we decide about disabling, we can do additional changes in that as require.

I've changed it and it works, at least locally for me :)

felixfontein committed:
Extend docker_container healthcheck (#1)

* Allowing to disable healthcheck.

* Added test for healthcheck.

* Make sure correct types are used.

* Healthcheck needs to be explicitly disabled with test: ['NONE'].

akshay196 committed:
pep8 fixes

Signed-off-by: Akshay Gaikwad <akgaikwad001@gmail.com>

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1330,6 +1374,77 @@ def _parse_log_config(self):
+                time_params[name] = int(value)
+
+        time = timedelta(**time_params)
+        time_in_nanoseconds = (time.seconds * 1000000 + time.microseconds) * 1000
```
Ok, there's one thing we forgot: a `timedelta` object also has `days`, and seconds are up to one day. Fortunately, there's an easier way to do this all:
```suggestion
        time_in_nanoseconds = int(time.total_seconds() * 1000000000)
```

felixfontein replied:
(Using `((time.days * 3600 * 24 + time.seconds) * 1000000 + time.microseconds) * 1000` is more exact, but that exactness is only relevant [if the time interval is more than 270 years or so](https://docs.python.org/3/library/datetime.html#datetime.timedelta.total_seconds), which seems *slightly* unreasonable for docker healthchecks ;-) )

akshay196 replied:
I agree with changes. Thanks @felixfontein for suggestion.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -2381,6 +2512,9 @@ def _parse_comparisons(self):
         comparisons['expected_ports'] = dict(type='dict', comparison=comparisons['published_ports']['comparison'], name='expected_ports')
+        comparisons['disable_healthcheck'] = dict(type='value',
+                                                  comparison='ignore' if comparisons['healthcheck']['comparison'] == 'ignore' else 'strict',
+                                                  name='disable_healthcheck')
```
Thanks for splitting this :) I forgot to run `flake8` manually...

felixfontein commented on test/integration/targets/docker_container/tasks/tests/options.yml:
```diff
@@ -1133,6 +1133,99 @@
+      retries: 3
+    stop_timeout: 1
+  register: healthcheck_3
+
```
There's one more test which could be relevant (since healthcheck parsing is pretty complex, it could happen that not specifying a healthcheck is treated as healthcheck changed or removed, which we want to avoid):
```suggestion

- name: healthcheck (no change)
  docker_container:
    image: alpine:3.8
    command: '/bin/sh -c "sleep 10m"'
    name: "{{ cname }}"
    state: started
    stop_timeout: 1
  register: healthcheck_4
```
(The `healthcheck_` numbers have to be adjusted below, and a `healthcheck_4 is not changed` has to be added to the assert.)

akshay196 replied:
Make sense. I will update healthcheck tests.

felixfontein reviewed (commented):
Looks good, up to two minor improvements (one bug if the healthcheck interval is 1 day or more, and one additional test).

akshay196 committed:
Fix bug if healthcheck interval is 1 day or more

`timedelta` object has days too and seconds are up to one day.
Therefore use `total_seconds()` to convert time into seconds.

Signed-off-by: Akshay Gaikwad <akgaikwad001@gmail.com>

akshay196 committed:
Add test for healthcheck when healthcheck is not specified

This is to avoid the situation when healthcheck is not specified and
treat this as healthcheck is changed or removed.

Signed-off-by: Akshay Gaikwad <akgaikwad001@gmail.com>

felixfontein commented:
The failed test is completely unrelated to the PR itself. I've asked on IRC if someone can restart it.

felixfontein reviewed (approved):
shipit

@kassiansun Can you also take a look at this one?

willthames commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -659,6 +674,29 @@
+      test: ["NONE"]
 '''
 
 RETURN = '''
```
Is RETURN not impacted by this change at all?

felixfontein replied:
No, it simply returns the equivalent of `docker inspect` as returned by the API, so it already returned healthcheck information (even though the module didn't allow to modify/set it, in case the container was created by something else with a healthcheck, that information was already available here).

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1330,6 +1374,77 @@ def _parse_log_config(self):
+
+        duration_options = ['interval', 'timeout', 'start_period']
+
+        for (key, value) in options.items():
```
Why not use a `list` as `key, value` are all same? `for option in options` seems make later logic clearer.

felixfontein replied:
I guess the only reason would be if we'd rename something in the future. But yeah, a list is probably better.

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1330,6 +1374,77 @@ def _parse_log_config(self):
+        for (key, value) in options.items():
+            if value in self.healthcheck:
+                if value in duration_options:
+                    time = self._convert_duration_to_nanosecond(self.healthcheck.get(value))
```
Should we pass it into the `result` if `time` is zero?

felixfontein replied:
Hmm, I think so. If the user specified zero, it probably has a good reason (or should fail if docker doesn't like it).

akshay196 replied:
When specify healthcheck parameter to zero, eg. `--health-interval=0s ` in following example
```
docker run --name=nginx-proxy -d --health-cmd='curl --fail http://nginx.host.com || exit 1' --health-interval=0s   nginx:1.13
```
Then check `docker inspect <container id>`, It does not displays that parameter.
```
...
            "Healthcheck": {
                "Test": [
                    "CMD-SHELL",
                    "curl --fail http://nginx.host.com || exit 1"
                ]
            },
```
 Does it mean docker daemon ignore such parameter?

kassiansun replied:
Ah, this is `golang`'s `omitempty`: https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/api/types/container/config.go#L28

felixfontein replied:
In that case, we'd also better drop it, otherwise we'll get idempotency problems...

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1330,6 +1374,77 @@ def _parse_log_config(self):
+                        result[key] = time
+                elif self.healthcheck.get(value):
+                    result[key] = self.healthcheck.get(value)
+                    if key == 'test':
```
We can reduce the level of indents:

```python
elif option == 'test':
elif option == 'retries':
else:
    result[option] = self.healthcheck.get(option)
```

And I don't thinks someone will be bored enough to specify a `test: ""` to get into the `elif self.healthcheck.get(value):
` check

felixfontein replied:
Hmm, that could improve it a bit. It will be still a bit complex, though :)

(Also, never underestimate users which will specify something which doesn't make sense ;) )

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1330,6 +1374,77 @@ def _parse_log_config(self):
+                    elif key == 'retries':
+                        try:
+                            result[key] = int(result[key])
+                        except Exception as e:
```
Should we detect such a stupid mistake from users xD?

felixfontein replied:
Yes. It could happen that someone templates something and gets the wrong value. If docker_container barfs with an exception, we get an Issue, if it returns a nice error, the user will hopefully solve the problem without filing an Issue ;)

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1330,6 +1374,77 @@ def _parse_log_config(self):
+                            self.fail('Cannot parse number of retries for healthcheck. '
+                                      'Expected an integer, got "{0}".'.format(result[key]))
+
+        if result['test'] == ['NONE']:
```
I think the `comparisons` can handle such kind of usage @felixfontein ? And per the documentation, `NONE` seems only used in `Dockerfile` for overriding the parent configuration.

felixfontein replied:
It's a bit more complicated. `test: ['NONE']` is equivalent to not having a healthcheck, which is something `comparsions` doesn't understand. We decided to use the `test: ['NONE']` syntax to disable healthchecks because the alternatives are kind of strange. Accepting `healthcheck: None` doesn't work (since that means "I don't care if the container has a healthcheck"), and `healthcheck: { }` (which we had before) is also kind of strange (and still needs the extra complexity with `disable_healthcheck` since the default comparison for dicts, `allow_more_present`, won't detect disabling in this case).

Having the explicit `test: ['NONE']` seems like the best way (and the most compatbile one with `comparisons`).

kassiansun replied:
I think `healthcheck: None` and `comparisons: heathcheck: strict` is enough for this?

Well, I prefer using existing functions.

felixfontein replied:
No, because `None` is the same as not specified, and thus `None` always means that no comparison is done. We really need something which is explicitly not `None`.

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -2434,11 +2568,17 @@ def __init__(self, **kwargs):
 
+        healthcheck_supported = LooseVersion(docker_version) >= LooseVersion('2.0')
+        if self.module.params.get("healthcheck") and not healthcheck_supported:
+            self.module.warn("docker or docker-py version is %s. Minimum version required is 2.0 to set healthcheck option. "
```
I think we should better fail here? It will be more consistent with other non-supported options.

felixfontein replied:
Yes, it should be `fail` instead of `warn`.

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -1584,6 +1701,8 @@ def has_different_configuration(self, image):
             working_dir=config.get('WorkingDir'),
             publish_all_ports=host_config.get('PublishAllPorts'),
+            expected_healthcheck=config.get('Healthcheck'),
+            disable_healthcheck=(not config.get('Healthcheck') or config.get('Healthcheck').get('Test') == ['NONE']),
```
I think the `disable_healthcheck` is redundant, we can leave the logic to `comparisons` and users to handle it. Introducing this logic might also conflict with `docker`'s original logic, as `docker run --health-cmd ANYTHING` will always generate a "CMD-SHELL" entry in `HealthCheck.Test`(tested on CentOS with docker 1.13).

felixfontein replied:
It doesn't really work that way, since docker indicates "no healthcheck" by returning `None` (I think), which isn't compatible to what `comparisons` does.

I don't understand why this logic (you mean `disable_healthcheck`?) conflicts with `docker`'s original logic?

In any way, I think we have to convert the string syntax for `test` to `CMD-SHELL`, otherwise idempotency won't work. That's a good point! (We should also have a test for it.)

kassiansun replied:
Yes, so I think we just need to `strict` compare these two options.

akshay196 replied:
> In any way, I think we have to convert the string syntax for test to CMD-SHELL

Right. There could be easy fix, like:
```
if isinstance(result[key], (tuple, list)):
    result[key] = [str(e) for e in result[key]]
else:
    result[key] = ["CMD-SHELL", str(result[key])]
```

I will update code for this along with test for it.

felixfontein replied:
Yes, that looks good. In particular, add an idempotency test for this one.

kassiansun commented:
BTW, I think there're too many `None` or some empty thing checks, I thinks this is against `python` best practice? I'm not familiar with `python`, at least with `ruby` we almost never do such kind of detection.

felixfontein commented:
> BTW, I think there're too many `None` or some empty thing checks, I thinks this is against `python` best practice? I'm not familiar with `python`, at least with `ruby` we almost never do such kind of detection.

I guess the reason is that in Ansible, and docker_container in particular, there's a big difference between `None` (which usually means "argument was not specified") and something else (which means something was specified). So there's always a need to distinguish between these cases.

akshay196 committed:
Convert string syntax for healthcheck test to CMD-SHELL

Also add another test case to check idempotency when healthcheck test
is specified as string

Signed-off-by: Akshay Gaikwad <akgaikwad001@gmail.com>

akshay196 committed:
Playbook fails if minimun docker version is not satisfy for healthcheck

This is to make more consistent with other non-supported options.

Signed-off-by: Akshay Gaikwad <akgaikwad001@gmail.com>

kassiansun commented:
shipit

felixfontein reviewed (approved):
shipit

felixfontein commented:
@akshay196 thanks a lot for working on this! This new feature will show up in Ansible 2.8.
@kassiansun thanks for reviewing!
@gundalow thanks for merging!

akshay196 commented:
@felixfontein @kassiansun @gundalow Thank you very much. I really appreciate your help. :)
